sound/soc/sof/sof-ipc-test-client.c

```diff
@@ -0,0 +1,329 @@
+ * The results of the last flood test can be accessed by reading the debugfs
+ * entries.
+ */
+static int sof_ipc_test_probe(struct auxiliary_device *auxdev,
+			      const struct auxiliary_device_id *id)
+{
+	struct sof_client_dev *cdev = auxiliary_dev_to_sof_client_dev(auxdev);
+	struct sof_ipc_client_data *ipc_client_data;
+
+	/*
+	 * The auxiliary device has a usage count of 0 even before runtime PM
+	 * is enabled. So, increment the usage count to let the device
+	 * suspend after probe is complete.
+	 */
+	pm_runtime_get_noresume(&auxdev->dev);
+
+	/* allocate memory for client data */
+	ipc_client_data = devm_kzalloc(&auxdev->dev, sizeof(*ipc_client_data), GFP_KERNEL);
+	if (!ipc_client_data)
+		return -ENOMEM;
+
+	ipc_client_data->buf = devm_kzalloc(&auxdev->dev, IPC_FLOOD_TEST_RESULT_LEN, GFP_KERNEL);
+	if (!ipc_client_data->buf)
+		return -ENOMEM;
+
+	cdev->data = ipc_client_data;
+
+	/* create debugfs root folder with device name under parent SOF dir */
+	ipc_client_data->dfs_root = debugfs_create_dir(dev_name(&auxdev->dev),
+						       sof_client_get_debugfs_root(cdev));
+
+	/* create read-write ipc_flood_count debugfs entry */
+	debugfs_create_file("ipc_flood_count", 0644, ipc_client_data->dfs_root,
+			    cdev, &sof_ipc_dfs_fops);
+
+	/* create read-write ipc_flood_duration_ms debugfs entry */
+	debugfs_create_file("ipc_flood_duration_ms", 0644, ipc_client_data->dfs_root,
+			    cdev, &sof_ipc_dfs_fops);
+
+	/* enable runtime PM */
+	pm_runtime_set_autosuspend_delay(&auxdev->dev, SOF_IPC_CLIENT_SUSPEND_DELAY_MS);
+	pm_runtime_use_autosuspend(&auxdev->dev);
+	pm_runtime_set_active(&auxdev->dev);
+	pm_runtime_enable(&auxdev->dev);
+	pm_runtime_mark_last_busy(&auxdev->dev);
+	pm_runtime_put_autosuspend(&auxdev->dev);
+
+	return 0;
+}
+
+static int sof_ipc_test_cleanup(struct auxiliary_device *auxdev)
+{
+	struct sof_client_dev *cdev = auxiliary_dev_to_sof_client_dev(auxdev);
+	struct sof_ipc_client_data *ipc_client_data = cdev->data;
+
+	pm_runtime_disable(&auxdev->dev);
+
+	debugfs_remove_recursive(ipc_client_data->dfs_root);
+
+	return 0;
+}
+
+static int sof_ipc_test_remove(struct auxiliary_device *auxdev)
+{
+	return sof_ipc_test_cleanup(auxdev);
+}
+
+static void sof_ipc_test_shutdown(struct auxiliary_device *auxdev)
+{
+	sof_ipc_test_cleanup(auxdev);
+}
+
+static const struct auxiliary_device_id sof_ipc_auxbus_id_table[] = {
+	{ .name = "snd_sof_client.ipc_test" },
+	{},
+};
+MODULE_DEVICE_TABLE(auxiliary, sof_ipc_auxbus_id_table);
+
+/*
+ * No need for driver pm_ops as the generic pm callbacks in the auxiliary bus type are enough to
+ * ensure that the parent SOF device resumes to bring the DSP back to D0.
+ * driver name will be set based on KBUILD_MODNAME.
+ */
+static struct sof_client_drv sof_ipc_test_client_drv = {
+	.auxiliary_drv = {
+		.id_table = sof_ipc_auxbus_id_table,
+		.probe = sof_ipc_test_probe,
+		.remove = sof_ipc_test_remove,
+		.shutdown = sof_ipc_test_shutdown,
+	},
```

Comment:
don't we need a name?

In the documentation, we state

.driver = {
				.name = AUX_DRV_NAME("myauxiliarydrv"),
			},

I did the same with SoundWire, not sure why  it's not done here?

Reply:
this part is the new change @plbossart . With KMOD_BUILDNAME, we've made name optional here. If provided, it will get appended with KMOD_BUILDNAME, if not, it will be KMOD_BUILDNAME. It was Leon's suggestion

Reply:
Thanks @ranj063, I will remove this name as well for the SoundWire stuff since I don't need it either.